File: CS/Form1.cs
using System;
using System.Data;
using System.Windows.Forms;
using DevExpress.XtraRichEdit;
using DevExpress.XtraRichEdit.API.Native;

namespace RichEditCalculatedField {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();

            richEditControl1.Options.MailMerge.DataSource = ProductsTable.CreateData();
            RestoreTemplate();
        }

        private void button1_Click(object sender, EventArgs e) {
            MailMergeOptions mailMergeOptions = richEditControl1.Document.CreateMailMergeOptions();
            mailMergeOptions.MergeMode = MergeMode.JoinTables;

            RichEditDocumentServer server = new RichEditDocumentServer();

            server.CalculateDocumentVariable += server_CalculateDocumentVariable;

            richEditControl1.Document.MailMerge(mailMergeOptions, server.Document);

            richEditControl1.LoadDocument("HeaderTemplate.rtf");
            richEditControl1.Document.AppendDocumentContent(server.Document.Range);
        }

        void server_CalculateDocumentVariable(object sender, CalculateDocumentVariableEventArgs e) {
            if (e.VariableName == "Prod") {
                int productId = -1;

                if (Int32.TryParse(e.Arguments[0].Value, out productId)) {
                    DataRow row = ((DataTable)richEditControl1.Options.MailMerge.DataSource).Rows.Find(productId);
                    int unitsInStock = Convert.ToInt32(row[e.Arguments[1].Value]);
                    decimal unitPrice = Convert.ToDecimal(row[e.Arguments[2].Value]);

                    e.Value = unitsInStock * unitPrice;
                    e.Handled = true;
                }
            }
        }

        #region Helper Methods
        private void button2_Click(object sender, EventArgs e) {
            RestoreTemplate();
        }

        void RestoreTemplate() {
            richEditControl1.LoadDocument("DetailTemplate.rtf");
            ShowFieldCodes();
        }

        private void ShowFieldCodes() {
            Document doc = richEditControl1.Document;
            doc.BeginUpdate();
            foreach (Field f in doc.Fields) f.ShowCodes = true;
            doc.EndUpdate();
        }
        #endregion Helper Methods
    }
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Data
Imports System.Windows.Forms
Imports DevExpress.XtraRichEdit
Imports DevExpress.XtraRichEdit.API.Native

Namespace RichEditCalculatedField
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()

			richEditControl1.Options.MailMerge.DataSource = ProductsTable.CreateData()
			RestoreTemplate()
		End Sub

		Private Sub button1_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click
			Dim mailMergeOptions As MailMergeOptions = richEditControl1.Document.CreateMailMergeOptions()
			mailMergeOptions.MergeMode = MergeMode.JoinTables

			Dim server As New RichEditDocumentServer()

			AddHandler server.CalculateDocumentVariable, AddressOf server_CalculateDocumentVariable

			richEditControl1.Document.MailMerge(mailMergeOptions, server.Document)

			richEditControl1.LoadDocument("HeaderTemplate.rtf")
			richEditControl1.Document.AppendDocumentContent(server.Document.Range)
		End Sub

		Private Sub server_CalculateDocumentVariable(ByVal sender As Object, ByVal e As CalculateDocumentVariableEventArgs)
			If e.VariableName = "Prod" Then
				Dim productId As Integer = -1

				If Int32.TryParse(e.Arguments(0).Value, productId) Then
					Dim row As DataRow = (CType(richEditControl1.Options.MailMerge.DataSource, DataTable)).Rows.Find(productId)
					Dim unitsInStock As Integer = Convert.ToInt32(row(e.Arguments(1).Value))
					Dim unitPrice As Decimal = Convert.ToDecimal(row(e.Arguments(2).Value))

					e.Value = unitsInStock * unitPrice
					e.Handled = True
				End If
			End If
		End Sub

		#Region "Helper Methods"
		Private Sub button2_Click(ByVal sender As Object, ByVal e As EventArgs) Handles button2.Click
			RestoreTemplate()
		End Sub

		Private Sub RestoreTemplate()
			richEditControl1.LoadDocument("DetailTemplate.rtf")
			ShowFieldCodes()
		End Sub

		Private Sub ShowFieldCodes()
			Dim doc As Document = richEditControl1.Document
			doc.BeginUpdate()
			For Each f As Field In doc.Fields
				f.ShowCodes = True
			Next f
			doc.EndUpdate()
		End Sub
		#End Region ' Helper Methods
	End Class
End Namespace